logger.go

```diff
@@ -28,33 +28,346 @@ import (
+		temp := [2]string{"invalid level ", strconv.Itoa(int(recv))}
+		return strings.Join(temp[:], "")
+	}
+}
+
+type LogField struct {
+	Name  string
+	Value LogFieldValue
+}
+
+func newLogField(name string, value LogFieldValue) LogField {
+	return LogField{
+		Name:  name,
+		Value: value,
+	}
 }
 
-func (l *testLogger) Print(v ...interface{})                 { fmt.Fprint(&l.capture, v...) }
-func (l *testLogger) Printf(format string, v ...interface{}) { fmt.Fprintf(&l.capture, format, v...) }
-func (l *testLogger) Println(v ...interface{})               { fmt.Fprintln(&l.capture, v...) }
-func (l *testLogger) String() string                         { return l.capture.String() }
+func newLogFieldIp(name string, value net.IP) LogField {
+	var str string
+	if value != nil {
+		str = value.String()
+	}
+	return newLogField(name, logFieldValueString(str))
+}
+
+func newLogFieldError(name string, value error) LogField {
+	var str string
+	if value != nil {
+		str = value.Error()
+	}
+	return newLogField(name, logFieldValueString(str))
+}
+
+func newLogFieldStringer(name string, value fmt.Stringer) LogField {
+	var str string
+	if value != nil {
+		str = value.String()
+	}
+	return newLogField(name, logFieldValueString(str))
+}
+
+func newLogFieldString(name string, value string) LogField {
+	return newLogField(name, logFieldValueString(value))
+}
+
+func newLogFieldInt(name string, value int) LogField {
+	return newLogField(name, logFieldValueInt64(int64(value)))
+}
+
+func newLogFieldBool(name string, value bool) LogField {
+	return newLogField(name, logFieldValueBool(value))
+}
+
+type StructuredLogger interface {
+	Error(msg string, fields ...LogField)
+	Warning(msg string, fields ...LogField)
+	Info(msg string, fields ...LogField)
+	Debug(msg string, fields ...LogField)
+}
+
+// A LogFieldValue can represent any Go value, but unlike type any,
+// it can represent most small values without an allocation.
+// The zero Value corresponds to nil.
+type LogFieldValue struct {
+	num uint64
+	any interface{}
+}
+
+// LogFieldValueType is the type of a [LogFieldValue].
+type LogFieldValueType int
+
+// It's important that LogFieldTypeAny is 0 so that a zero Value represents nil.
+const (
+	LogFieldTypeAny LogFieldValueType = iota
+	LogFieldTypeBool
+	LogFieldTypeInt64
+	LogFieldTypeString
+)
+
+// LogFieldValueType returns v's LogFieldValueType.
+func (v LogFieldValue) LogFieldValueType() LogFieldValueType {
+	switch x := v.any.(type) {
+	case LogFieldValueType:
+		return x
+	case string:
+		return LogFieldTypeString
+	default:
+		return LogFieldTypeAny
+	}
+}
+
+func logFieldValueString(value string) LogFieldValue {
+	return LogFieldValue{num: uint64(len(value)), any: value}
+}
 
-type defaultLogger struct{}
+func logFieldValueInt(v int) LogFieldValue {
+	return logFieldValueInt64(int64(v))
+}
+
+func logFieldValueInt64(v int64) LogFieldValue {
+	return LogFieldValue{num: uint64(v), any: LogFieldTypeInt64}
```

Comment:
rather than setting `any` to a particular type, would it be easier to just have a separate field on the struct of the type?